crates/oxc_isolated_declarations/src/lib.rs

```diff
@@ -77,6 +81,10 @@ impl<'a> IsolatedDeclarations<'a> {
         &mut self,
         program: &Program<'a>,
     ) -> oxc_allocator::Vec<'a, Statement<'a>> {
+        // Collect information about global Symbol usage within computed
+        // properties before performing any transformations.
+        self.global_symbol_binding_tracker.visit_program(program);
```

Comment:
We could probably avoid allocations for this most of the time by first checking if there is both an occurrence of overriding the global Symbol and references to Symbol *anywhere* in the program, not necessarily in the same scope, but I'm not sure what the cost of that extra traversal would be.

Reply:
Is it possible to check on demand? For example, we initially only check for the root scope. When we visit the function we will check the function. This will only check for declarations that will be exported

Reply:
Possibly, I will think about it some more. I think one of the main challenges comes just from organizing the code and performing state management. It feels like similar to with `ScopeTree` we would need to tell `GlobalSymbolBindingTracker` when it is entering or leaving a scope, but the logic will diverge from `ScopeTree` (otherwise we would have just used `ScopeTree` for this purpose), which may result in the code looking more confusing.

(Just voicing my thoughts out loud) I think what feels redundant with my current approach is that the short-circuit optimizations I take are making assumptions about how the tree will be traversed for emitting declarations, so that logic kind of feels like it's being reproduced in two places. The code could be potentially cleaner if that logic could be pulled out. What I imagine is something like alternative `Visitor` traits that only traverse the parts of the AST relevant to emitted declarations. Such traits would already have the "optimizations" I wrote naturally built-in. What's harder for me to imagine right now is how easy it would be to re-implement declaration transforms on top of such traits.

Please let me know your thoughts.

Reply:
> Possibly, I will think about it some more. I think one of the main challenges comes just from organizing the code and performing state management. It feels like similar to with `ScopeTree` we would need to tell `GlobalSymbolBindingTracker` when it is entering or leaving a scope, but the logic will diverge from `ScopeTree` (otherwise we would have just used `ScopeTree` for this purpose), which may result in the code looking more confusing.
> 

Only `Function` and `TSModuleDeclaration` will have nested scopes, which I don't think should be too much trouble. We could try.

> (Just voicing my thoughts out loud) I think what feels redundant with my current approach is that the short-circuit optimizations I take are making assumptions about how the tree will be traversed for emitting declarations, so that logic kind of feels like it's being reproduced in two places. The code could be potentially cleaner if that logic could be pulled out. What I imagine is something like alternative `Visitor` traits that only traverse the parts of the AST relevant to emitted declarations. Such traits would already have the "optimizations" I wrote naturally built-in. What's harder for me to imagine right now is how easy it would be to re-implement declaration transforms on top of such traits.
> 

Adding another trait instead of `Visitor` may significantly increase maintenance costs. Currently `Visitor` is automatically generated with `ast-codegen`

Reply:
> Adding another trait instead of Visitor may significantly increase maintenance costs

That's a good point. I suppose it would be necessary for such a trait to extend `Visit` and only override a small subset of the method implementations.

> Only `Function` and `TSModuleDeclaration` will have nested scopes, which I don't think should be too much trouble. We could try.

I may not be able to get to this soon, but as you said, we have time!